Source/Parallelization/WarpXComm_K.H

```diff
@@ -641,118 +486,145 @@ void warpx_interp_nd_efield_z (int j, int k, int l,
-    }
-    Eza(j,k,l) = res;
-#else
-    amrex::Real res = 0.;
-    for (int nz = 1; nz < noz/2+1; nz++) {
-        res += 0.5_rt * stencil_coef_z[nz-1] * (Ezf(j, k, l + nz - 1) + Ezf(j, k, l - nz));
-    }
-    Eza(j,k,l) = res;
+    // Additional normalization factor
+    const amrex::Real wj = (sj == 0) ? 0.5_rt : 1.0_rt;
+    const amrex::Real wk = (sk == 0) ? 0.5_rt : 1.0_rt;
+#if   (AMREX_SPACEDIM == 2)
+    constexpr amrex::Real wl = 1.0_rt;
+#elif (AMREX_SPACEDIM == 3)
+    const amrex::Real wl = (sl == 0) ? 0.5_rt : 1.0_rt;
+#endif
+
+    // Arbitrary-order stencil coefficients
+    // (with PSATD they are reassigned in the interpolation loop below)
+    amrex::Real cj = 1.0_rt;
+    amrex::Real ck = 1.0_rt;
+    amrex::Real cl = 1.0_rt;
+
+    // Min and max for interpolation loop along j
+    const int jmin = (sj == 0) ? j - nj + 1 : j;
+    const int jmax = (sj == 0) ? j + nj - 2 : j;
+
+    // Min and max for interpolation loop along k
+    const int kmin = (sk == 0) ? k - nk + 1 : k;
+    const int kmax = (sk == 0) ? k + nk - 2 : k;
+
+    // Min and max for interpolation loop along l
+#if   (AMREX_SPACEDIM == 2)
+    // l = 0 always
+    const int lmin = l;
+    const int lmax = l;
+#elif (AMREX_SPACEDIM == 3)
+    const int lmin = (sl == 0) ? l - nl + 1 : l;
+    const int lmax = (sl == 0) ? l + nl - 2 : l;
 #endif
+
+    // Integer indices to access arbitrary-order stencil coefficients
+    int dj, dk;
+#if (AMREX_SPACEDIM == 3)
+    int dl;
+#endif
+
+    // Interpolation loop (see below examples and illustrations for 1D case).
+    // So far, with FDTD we use only interpolation order 2 and we do not have access
+    // to the Fornberg coefficients. In the future, we might want to use the Fornberg
+    // coefficients with both FDTD and PSATD, in which case the #ifdef WARPX_USE_PSATD condition
+    // below will be removed. Currently, with FDTD we set cj = ck = cl = 1 and do not change them.
+    //
+    // - example and illustration: nodal to nodal
+    //
+    //           j
+    //   --o-----o-----o--  result(j) = f(j)
+    //   --o-----o-----o--
+    //    j-1    j    j+1
+    //
+    // - example and illustration: staggered to nodal
+    //
+    //                       j
+    //   --o-----o-----o-----o-----o-----o-----o--  result(j) = c_0 * (f(j-1) + f(j)  ) / 2
+    //   -----x-----x-----x-----x-----x-----x-----            + c_1 * (f(j-2) + f(j+1)) / 2
+    //       j-3   j-2   j-1    j    j+1   j+2                + c_2 * (f(j-3) + f(j+2)) / 2
+    //       c_2   c_1   c_0   c_0   c_1   c_2                + ...
+    //
+    amrex::Real res = 0.0_rt;
+    for         (int ll = lmin; ll <= lmax; ll++) {
+        for     (int kk = kmin; kk <= kmax; kk++) {
+            for (int jj = jmin; jj <= jmax; jj++) {
+#ifdef WARPX_USE_PSATD
+                dj = (j - jj > 0) ? j - jj - 1 : jj - j;
+                dk = (k - kk > 0) ? k - kk - 1 : kk - k;
+                if (sj == 0) cj = stencil_coef_x[dj];
```

Comment:
We found from experience that using the tenary operator vectorizes better:
```suggestion
                cj = (sj == 0) ? stencil_coef_x[dj];
```

We can check on https://cuda.godbolt.org what happens on the low level and if the cost for `sj != 0` are severe, we can add some template magic and compile the kernel twice w/o much code duplication.